File: gtci/18-bitwise-xor/python3/2-single-number.py
class Solution:
    def findSingleNumber(self, arr):
        num = arr[0]

        for i in range(1, len(arr)):
            num ^= arr[i]

        return num


if __name__ == "__main__":
    sol = Solution()

    assert sol.findSingleNumber([2, 2, 1]) == 1

    assert sol.findSingleNumber([4, 1, 2, 1, 2]) == 4

    assert sol.findSingleNumber([1]) == 1

    print("All test cases passed.")

File: gtci/18-bitwise-xor/cpp/2-single-number.cpp
#include <cassert>
#include <iostream>
#include <vector>

class Solution {
  public:
    int findSingleNumber(const std::vector<int>& arr) {
      int num = arr[0];
      
      for (size_t i = 1; i < arr.size(); i++)
        num ^= arr[i];

      return num;
    }
};

int main() {
    Solution s;

    assert(s.findSingleNumber({2, 3, 5, 4, 5, 3, 4}) == 2);
    assert(s.findSingleNumber({1, 1, 2}) == 2);
    assert(s.findSingleNumber({1}) == 1);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}
